### ISSUE TITLE
how to add include path for language servers in coc-settings.json file?

### ISSUE THREAD

**Issue by amirrezasadeqi** [clangd/coc-clangd#28]
Hi there.

since I work in ros platform I want to add the headers of the ros(/opt/ros/kinetic/include/**) to my default include path of language servers for ros projects. with out this option, at the beginning of each project I must add a dummy ros node(is ros node means a cpp file), and compile with `catkin_make -DCMAKE_EXPORT_COMPILE_COMMANDS=YES` to generate the compile_commands.json file to get access to ros headers in the completion list.

the contents of my coc-setting file, is as below:

    {
    
    "languageserver": {
        "clangd": {
          "command": "clangd",
          "rootPatterns": ["compile_flags.txt", "compile_commands.json"],
          "filetypes": ["c", "cpp", "objc", "objcpp"]
        },
        "cquery": {
          "command": "/home/amirrezasadeghi/ManBuildLibs/cquery/build/cquery",
          "args": ["--log-file=/tmp/cq.log"],
          "filetypes": ["c", "cpp"],
          "rootPatterns": ["compile_flags.txt", "compile_commands.json", ".git/", ".hg/"],
          "initializationOptions": {
            "cacheDirectory": "/tmp/cquery"
          }
        },
        "ccls": {
          "command": "ccls",
          "filetypes": ["c", "cpp", "objc", "objcpp"],
          "rootPatterns": [".ccls", "compile_commands.json", ".git/", ".hg/"],
          "initializationOptions": {
             "cache": {
               "directory": "/tmp/ccls"
             }
           }
        }
      },
      "clangd.semanticHighlighting": true
    }

So my major question is how to add ros include path to these settings. I checked the [wiki page of coc.nvim][2] but can't find any option for user include path. also, I checked the option of clangd and tried to add, 
`"args":["-I/opt/ros/kinetic/include/"]`,
to the corresponding block but it did not work. also, I added [.cquery][3] file, but still no effect on the list of the proposed headers.

sorry for long story, and thanks for your time.

Edit: it would be better to use the clangd LSP rather than ccls. because it is faster and has better documentation. thanks!

  [1]: https://github.com/neoclide/coc.nvim
  [2]: https://github.com/neoclide/coc.nvim/wiki
  [3]: https://github.com/cquery-project/cquery/wiki/.cquery

**Comment by fannheyward** [comment]
1. Why are you mix-use clangd/cquery/ccls?
2. When you use `coc-clangd`, you should not use `languageserver` to configuration clangd in `coc-settings.json`
3. Maybe you need to set `compile_commands.json` to include what you want.

**Comment by sam-mccall** [comment]
Generally speaking clangd treats include paths as a per-project setting, and you should have a compile_commands.json file generated by your build system, or the simpler compile_flags.txt which could contain just `-I/opt/ros/kinetic/include`. http://clangd.llvm.org/installation.html#project-setup

If you don't want to add this to every project directory, you can place the compile_flags.txt somewhere (`/some/for/compile_flags.txt`) and use the clangd flag `-compile-commands-dir=/some/dir`.

Be aware that using compile_flags.txt means background indexing won't work, as it doesn't provide the list of files to consider part of the project.

**Comment by amirrezasadeqi** [comment]
I don't use all of the LSPs at the same time just to be complete and I disable them base on the project.

as I said, I use the ros platform. the build system of ros projects is CMake and it uses a tool which is called catkin_make. catkin_make creates a CMakeLists.txt file in the top-level src directory([details of the ros project tree](http://wiki.ros.org/ROS/Tutorials/CreatingPackage)) and handles the content of this file automatically. each package that I create has its own CMakeLists.txt file and catkin_make automatically combines CMake files of packages in the top-level CMake file.

so if I want to have ros headers in completion suggestions, I must create first a package and edit its CMake file and rebuild the project using catkin_make(i think it automatically adds the ros directories to the top-level CMake file and so the generated compile_commands.json would contain "-I" option for inclusion of ros headers.)

catkin_make also creates a hidden file `.catkin_workspace` in the root directory. VSCode uses this file to determine if the project is a ros project and also the root of the project. so in VSCode I just create a project and without adding any package, I build it with catkin_make and when I open any cpp file in that project in VSCode the ros headers will pop out in suggestions.

**THE MAJOR PROBLEM:**

so what actually I want is: 

1. automate this procedure like what VSCode does.

2. VSCode has a c_cpp_properties.json file, which contains a feature "includePath". I can add any path to this feature to include in completion suggestions. what is the corresponding feature in clangd?

3. can I use `compile_flags.txt` and `compile_commands.json` at the same time? and does coc-clangd combine these files?

excuse me for long text. I wanted to explain everything. thanks a lot.

**Comment by amirrezasadeqi** [comment]
excuse me fann. I don't know exactly what 'help wanted' tag means. do you need more details? and if yes, about which part?

**Comment by fannheyward** [comment]
@amirrezasadeqi  `c_cpp_properties.json` in VSCode is used by [C/C++](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools), not clangd, I don't known whether clangd support this or not.

`coc-clangd` is an extension for clangd, what it can do depends on what clangd can do.

@sam-mccall does clangd supports this?

**Comment by amirrezasadeqi** [comment]
@fannheyward do you mean that C/C++ is like a language server which is different from clangd?
if this is true, I think there is no [feature in clangd](https://manpages.debian.org/experimental/clangd/clangd.1.en.html) to support the includePath, in the way vscode does.

So to support this feature, we must wait for clangd!!?

**Comment by fannheyward** [comment]
Yes, C/C++ does not use clangd AFAIK.

**Comment by amirrezasadeqi** [comment]
thank you guys for your help.

I have no experience in writing plugins for vim but that would be my pleasure if I can add this feature to the coc-setting (**if some day you find out there is possibility of implementation of this feature in clangd**).

**Comment by sam-mccall** [comment]
Clangd is configured by compile_commands.json or compile_flags.txt, but not both. It does not use c_cpp_properties.json.

CMake has support for generating compile_commands.json via the option CMAKE_EXPORT_COMPILE_COMMANDS. catkin_make is supposed to pass this value on if you set it as described here: https://github.com/catkin/catkin_tools/issues/551. It's unclear to me whether this will generate one top-level compile_commands.json or one per subproject. Either should work fine, you may have to symlink them from the build tree to the corresponding source tree if they're not the same.

coc-clangd is not a suitable place to add more automation around these workflows. clangd itself may be one day, but it would need a concrete proposal from someone familiar with the code. Meanwhile if you need more automation, I suggest you write a shell script to run cmake/catkin_* with the right flags and symlink compile_commands.json into the right place.

**Comment by amirrezasadeqi** [comment]
Hi there. thank you guys for your help and descriptions. finally, I found a way to make the procedure of adding include path like the procedure of VSCode. so I wrote a remote plugin in python([neovim_includePlug_draft.py.zip](https://github.com/clangd/coc-clangd/files/4490104/neovim_includePlug_draft.py.zip)) which exports a Command(`:IncludePath`). by running this command, it prompts you to enter the include path then it automatically adds this path to compile options of the corresponding file in the `compile_commands.json` file. so without regeneration of the JSON file, you will have headers of the included path in your autocompletion.
thank you all ...
